### ISSUE TITLE
Multithreading not working

### ISSUE THREAD

**Issue by baztian** [jpype-project/jpype#211]
Thought #210 should fix #203 as well so I ran a small test baztian@b9469c87f0a819134fc6f409c114c1514f6977c4  (using python 2.7.12). The results were a bit weird though.
I'm getting the following error if I run the complete testsuite.

    # A fatal error has been detected by the Java Runtime Environment:
    #
    #  SIGSEGV (0xb) at pc=0x00007ff32ec7b079, pid=9746, tid=0x00007ff319573700
    #
    # JRE version: Java(TM) SE Runtime Environment (8.0_111-b14) (build 1.8.0_111-b14)
    # Java VM: Java HotSpot(TM) 64-Bit Server VM (25.111-b14 mixed mode linux-amd64 compressed oops)
    # Problematic frame:
    # C  [_jpype.so+0x29079][thread 140682070533888 also had an error]
      JPJavaEnv::NewString(unsigned short const*, int)+0x29
    #
    # Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
    #
    # An error report file with more information is saved as:
    # /home/bbowe/Sources/jpype/hs_err_pid9746.log
    #
    # If you would like to submit a bug report, please visit:
    #   http://bugreport.java.com/bugreport/crash.jsp
    #
    Aborted

The test hangs forever if I run the test directly via `python test/testsuite.py jpypetest.threads`.
The test almost always runs if I uncomment the attachThreadToJVM/detachThreadFromJVM calls. But one time I got the following error:

    INFO:jpypetest.common:Running testsuite using JVM /usr/lib/jvm/java-8-oracle/jre/lib/amd64/server/libjvm.so
    #
    # A fatal error has been detected by the Java Runtime Environment:
    #
    #  SIGSEGV (0xb) at pc=0x00007f54dac61a24, pid=9853, tid=0x00007f54c1696700
    #
    # JRE version: Java(TM) SE Runtime Environment (8.0_111-b14) (build 1.8.0_111-b14)
    # Java VM: Java HotSpot(TM) 64-Bit Server VM (25.111-b14 mixed mode linux-amd64 compressed oops)
    # Problematic frame:
    # C  [_jpype.so+0x43a24]  JPMethod::findOverload(std::vector<HostRef*, std::allocator<HostRef*> >&, bool)+0x84
    #
    # Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
    #
    # An error report file with more information is saved as:
    # /home/bbowe/Sources/jpype/hs_err_pid9853.log
    #
    # If you would like to submit a bug report, please visit:
    #   http://bugreport.java.com/bugreport/crash.jsp
    # The crash happened outside the Java Virtual Machine in native code.
    # See problematic frame for where to report the bug.
    #
    Aborted

**Comment by marscher** [comment]
I think the caching mechanism for methods is not thread safe. Thanks for 
adding the test case. If we switch to a recent C++ standard we can 
easily guard the sections with mutexes.

**Comment by marscher** [comment]
@baztian could you please test https://github.com/marscher/jpype/tree/thread_safety?

**Comment by marscher** [comment]
This solution essentially forbids multiple threads to access the C++ API at the same time, because the API itself is not written in a thread safe manner and I guess it would be a lot of effort to do so. So we just need to synchronize the calls to it, which is a lot easier.

**Comment by baztian** [comment]
Thanks @marscher. It seems to work!
I've [added python 2 compatibility to the test and updated the documentation](/baztian/jpype/tree/thread_safety). Would be nice if you could merge all those changes and release it.

**Comment by marscher** [comment]
It is not yet functional. I adopted the testsuite to execute all test-cases in a threaded environment and there are still errors to fix. Sorry for the delay.

**Comment by mikea1729** [comment]
I didn't want to confuse all the existing work on getting multi-threading working well, so pushed a small PR #217 which at least fixes my problems.  And instead of creating a new issue, I'm posting here.

Tested on v0.6.2 and on master (07845b0).  I saw what may be similar crashes to those documented here.  A good test case is loading up ipython2 and then
```
from jpype import *
java.lang.System.out.
```
then press tab before enter and watch ipython try to auto-complete and hit a NULL pointer deref.  It seems that ipython (v 5.3.0) makes lots of LWP theads while handling console input so happens to be a good test for these threading issues.

**Comment by samuelrohr** [comment]
Hello, i'm having the same problem that @baztian commeted 

#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f88898fa609, pid=4356, tid=0x00007f886204d700
#
# JRE version: OpenJDK Runtime Environment (8.0_151-b12) (build 1.8.0_151-8u151-b12-0ubuntu0.17.10.2-b12)
# Java VM: OpenJDK 64-Bit Server VM (25.151-b12 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [_jpype.cpython-36m-x86_64-linux-gnu.so+0x37609]  JPJavaEnv::NewString(unsigned short const*, int)+0x29
#
# Core dump written. Default location: /home/samuel/Projetos/Django/recognition_service/core or core.4356
#
# An error report file with more information is saved as:
# /home/samuel/Projetos/Django/recognition_service/hs_err_pid4356.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#

What i have done:
1) Compiled the java project
2) Create a python module that is a interface for java methods. -> this works well
3) Create a python project that import the module created and use it (just for test) -> this works well too
4) Import the module (step 2) into a Django project. Here is where the things go wrong. I expose a url that call's my python module and this should call the java, but when access i got the error.

I'm using the jpype v0.6.2 but im not sure if this is the version with the thread_safe fix

**Comment by samuelrohr** [comment]
I found the solution on this link: https://stackoverflow.com/questions/7861299/jpype-doesnt-work-inside-thread

The problem was i need to attachThreadToJVM() before call the java method on my python module

**Comment by mikea1729** [comment]
Hi @samuelrohr.  Just an FYI, the fix you found may not be sufficient for your Django use case.  If you get lots of clients connecting in many threads, the fact that you called attachThreadToJVM() on the Python side could still put you in a race condition with another thread.  If another thread attaches you will hit the the NULL JNIEnv* bug my (unmerged) PR #217 was meant to solve.  If you end up needing it you'll probably want to rebase it against master as I haven't looked at it since May.

**Comment by samuelrohr** [comment]
@mikea1729 thx for the tip. I will look into it replacing the current jpype to yours and testing.

**Comment by Thrameos** [comment]
Threads are attached automatically in the current version.  It is still necessary to detach them to prevent memory leaks, but this case should no longer occur.
